tensorflow/contrib/rnn/kernels/lstm_ops.h

```diff
@@ -0,0 +1,272 @@
+    Eigen::array<int, 2> h_offsets  = {0, cell_size * 6};
+
+    Eigen::array<int, 2> cell_extents = {batch_size, cell_size};
+
+    // xh = [x, h_prev]
+    Eigen::array<int, 2> xh_x_offsets = {0, 0};
+    Eigen::array<int, 2> xh_x_extents = {batch_size, input_size};
+    Eigen::array<int, 2> xh_h_offsets = {0, input_size};
+    Eigen::array<int, 2> xh_h_extents = {batch_size, cell_size};
+    xh.slice(xh_x_offsets, xh_x_extents).device(d) = x;
+    auto h_prev = states_prev.slice(h_offsets, cell_extents);
+    xh.slice(xh_h_offsets, xh_h_extents).device(d) = h_prev;
+
+    // dh.
+    auto dh = states_prev_grad.slice(h_offsets, cell_extents);
+    dh.device(d) =
+        h_grad + states_grad.slice(h_offsets, cell_extents);
+
+    // do[t] = sigm'(o[t]) .* dh[t] .* co[t]
+    auto co = states.slice(co_offsets, cell_extents);
+    auto o = states.slice(o_offsets, cell_extents);
+    states_prev_grad.slice(o_offsets, cell_extents).device(d) =
+        o * (o.constant(1.0f) - o) * dh * co;
+
+    // dcs[t] += tanh'(cs[t]) .* dh[t] .* o[t] + dcs[t + 1] .* f[t + 1]
+    auto dcs = states_prev_grad.slice(co_offsets, cell_extents);
+    dcs.device(d) =
+        (co.constant(1.0f) - co * co) * dh * o +
+        states_grad.slice(ci_offsets, cell_extents);
+
+    // dci[t] = tanh'(ci[t]) dcs[t] i[t]
+    auto ci = states.slice(ci_offsets, cell_extents);
+    auto i = states.slice(i_offsets, cell_extents);
+    states_prev_grad.slice(cs_offsets, cell_extents).device(d) =
+        (ci.constant(1.0f) - ci * ci) * dcs * i;
+
+    // df[t] = sigm'(f[t]) dcs[t] cs[t - 1]
+    auto f = states.slice(f_offsets, cell_extents);
+    auto cs_prev = states_prev.slice(cs_offsets, cell_extents);
+    states_prev_grad.slice(f_offsets, cell_extents).device(d) =
+        f * (f.constant(1.0f) - f) * dcs * cs_prev;
+
+    // di[t] = sigm'(i[t]) dcs[t] ci[t]
+    states_prev_grad.slice(i_offsets, cell_extents).device(d) =
+        i * (i.constant(1.0f) - i) * dcs * ci;
+
+    // xh_grad.
+    Eigen::array<int, 2> states_offsets = {0, 0};
+    Eigen::array<int, 2> states_extents = {batch_size, cell_size * 4};
+
+    auto dstate4 = states_prev_grad.slice(states_offsets, states_extents);
+
+    Eigen::array<Eigen::IndexPair<Eigen::DenseIndex>, 1> xh_grad_contract_pairs;
+    xh_grad_contract_pairs[0] = Eigen::IndexPair<Eigen::DenseIndex>(1, 1);
+
+    // xh_grad = dstate4 * w^T
+    if (USE_CUBLAS) {
+#if GOOGLE_CUDA
+      auto kTranspose =
+          perftools::gputools::blas::Transpose::kTranspose;
+      auto kNoTranspose =
+          perftools::gputools::blas::Transpose::kNoTranspose;
+
+      const uint64 m = batch_size;
+      const uint64 k = cell_size * 4;
+      const uint64 n = input_size + cell_size;
+
+      auto a_ptr = AsDeviceMemory(states_prev_grad.data());
+      auto b_ptr = AsDeviceMemory(w.data());
+      auto c_ptr = AsDeviceMemory(xh_grad.data());
+
+      CHECK(CHECK_NOTNULL(stream)->ThenBlasGemm(
```

Comment:
can you pass context into this function and then set the status and return ?  Killing the whole program here is probably a bad idea when the error message can be sent back to the client.